lib/SIL/SILBasicBlock.cpp

```diff
@@ -187,9 +187,32 @@ transferNodesFromList(llvm::ilist_traits<SILBasicBlock> &SrcTraits,
   // If splicing blocks not in the same function, update the parent pointers.
   for (; First != Last; ++First) {
     First->Parent = Parent;
-    for (auto &II : *First)
+    for (auto &II : *First) {
       II.setDebugScope(B,
                        ScopeCloner.getOrCreateClonedScope(II.getDebugScope()));
+      SILInstruction *I = &II;
+      // Maintain the registrations of opened archetypes.
+      if (isa<OpenExistentialAddrInst>(I) || isa<OpenExistentialRefInst>(I) ||
+          isa<OpenExistentialBoxInst>(I)) {
+        auto Ty = I->getType().getSwiftRValueType();
+        if (isa<ArchetypeType>(Ty) &&
+            !cast<ArchetypeType>(Ty)->getOpenedExistentialType().isNull()) {
+          SrcTraits.Parent->removeOpenedArchetypeDef(Ty, I);
+          Parent->addOpenedArchetypeDef(Ty, I);
+        }
+      } else if (isa<OpenExistentialMetatypeInst>(I)) {
+        SILType InstanceTy = I->getType();
+        while (isa<AnyMetatypeType>(InstanceTy.getSwiftRValueType())) {
+          InstanceTy = InstanceTy.getMetatypeInstanceType(I->getModule());
+        }
+        auto Ty = InstanceTy.getSwiftRValueType();
+        if (isa<ArchetypeType>(Ty) &&
+            !cast<ArchetypeType>(Ty)->getOpenedExistentialType().isNull()) {
+          SrcTraits.Parent->removeOpenedArchetypeDef(Ty, I);
+          Parent->addOpenedArchetypeDef(Ty, I);
+        }
+      }
+    }
```

Comment:
It's basically every instruction that stores a Type or a Substitution.  I don't understand the need for it on function type arguments, though; it's not like inlining would just forward the type argument "uses", since any particular substituted type might be concrete, redundant, or involve multiple archetypes.

Eric: the only problem I see with putting it on the SILModule is that it means we have to update the structure when deleting a function body.